File: src/backend/models/canvas.py
from typing import Literal
from pydantic import BaseModel, Field

HexColor = str

# Shared contract
class PixelChange(BaseModel):
    x: int = Field(ge=0)
    y: int = Field(ge=0)
    color: HexColor

# User Sent
class PixelUpdate(BaseModel):
    type: Literal["pixel_update"]
    x: int = Field(ge=0)
    y: int = Field(ge=0)
    color: HexColor

class PixelBatchUpdate(BaseModel):
    type: Literal["pixel_batch_update"]
    pixels: list[PixelChange] = Field(min_length=1)
    
class PixelBatchUpdated(BaseModel):
    type: Literal["pixel_batch_updated"] = "pixel_batch_updated"
    pixels: list[PixelChange]

# Server Sent
class PixelUpdated(BaseModel):
    type: Literal["pixel_updated"] = "pixel_updated"
    x: int
    y: int
    color: HexColor

class CanvasSnapshot(BaseModel):
    width: int
    height: int
    defaultColor: HexColor
    pixels: list[list[HexColor]]

class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    message: str

File: src/ux/src/types/canvas.ts
export type HexColor = string;

export type PixelGrid = HexColor[][];

export interface CanvasSnapshot {
  width: number;
  height: number;
  defaultColor: HexColor;
  pixels: PixelGrid;
}

export interface PixelChange {
  x: number;
  y: number;
  color: HexColor;
}

export interface PixelUpdateMessage {
  type: "pixel_update";
  x: number;
  y: number;
  color: HexColor;
}

export interface PixelUpdatedMessage {
  type: "pixel_updated";
  x: number;
  y: number;
  color: HexColor;
}

export interface PixelBatchUpdateMessage {
  type: "pixel_batch_update";
  pixels: PixelChange[];
}

export interface PixelBatchUpdatedMessage {
  type: "pixel_batch_updated";
  pixels: PixelChange[];
}

export interface ErrorMessage {
  type: "error";
  message: string;
}

export type ServerCanvasMessage = PixelUpdatedMessage | PixelBatchUpdatedMessage | ErrorMessage;